### PR TITLE
Generate tests in a temporary directory instead of in-tree

### LINKED ISSUE #499: Tests should not generate files in the tests directory
TMPDIR is probably a better place than
/usr/share/cargo/registry/cbindgen-0.13.2/tests/expectations/typedef.compat.o (as example

Note that this is failing in Debian pacakges test CI:
https://ci.debian.net/packages/r/rust-cbindgen/

Example:
```
---- test_typedef stdout ----
Running: "/tmp/tmp.DS1uKzckhx/target/x86_64-unknown-linux-gnu/debug/cbindgen" "--lang" "c" "--cpp-compat" "--style" "type" "-o" "/usr/share/cargo/registry/cbindgen-0.13.2/tests/expectations/typedef.compat.c" "/usr/share/cargo/registry/cbindgen-0.13.2/tests/rust/typedef.rs"
Running: "gcc" "-D" "DEFINED" "-c" "/usr/share/cargo/registry/cbindgen-0.13.2/tests/expectations/typedef.compat.c" "-o" "/usr/share/cargo/registry/cbindgen-0.13.2/tests/expectations/typedef.compat.o" "-I" "/usr/share/cargo/registry/cbindgen-0.13.2/tests" "-Wall" "-Werror" "-Wno-attributes" "-g" "-O2" "-fdebug-prefix-map=/usr/share/cargo/registry/cbindgen-0.13.2=." "-fstack-protector-strong" "-Wformat" "-Werror=format-security" "-D" "CBINDGEN_STYLE_TYPE"
thread 'test_typedef' panicked at 'Output failed to compile: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "Assembler messages:\nFatal error: can\'t create /usr/share/cargo/registry/cbindgen-0.13.2/tests/expectations/typedef.compat.o: Permission denied\n" }', tests/tests.rs:106:5
stack backtrace:
   0: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
   1: core::fmt::write
   2: std::io::Write::write_fmt
   3: std::io::impls::<impl std::io::Write for alloc::boxed::Box<W>>::write_fmt
   4: std::panicking::default_hook::{{closure}}
   5: std::panicking::default_hook
   6: std::panicking::rust_panic_with_hook
   7: std::panicking::continue_panic_fmt
   8: std::panicking::begin_panic_fmt
   9: tests::compile
             at tests/tests.rs:106
  10: tests::run_compile_test
             at tests/tests.rs:150
  11: tests::test_file
             at tests/tests.rs:161
  12: tests::test_typedef
             at tests/tests.rs:185
  13: tests::test_typedef::{{closure}}
             at tests/tests.rs:184
  14: core::ops::function::FnOnce::call_once
             at /usr/src/rustc-1.40.0/src/libcore/ops/function.rs:227
  15: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
  16: __rust_maybe_catch_panic
  17: test::run_test_in_process
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

### PR DESCRIPTION
**Pull request by sylvestre**
Closes #499

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -23,6 +23,7 @@ tempfile = "3.0"
 toml = "0.5"
 proc-macro2 = "1"
 quote = "1"
+tempdir = "0.3"
```
Comment (emilio) [inline Cargo.toml:26]: This should be in the `[dev-dependencies]` section.

File: tests/tests.rs [hunk 2]
```diff
@@ -21,6 +22,10 @@ fn run_cbindgen(
     cpp_compat: bool,
     style: Option<Style>,
 ) {
+
```
Comment (emilio) [inline tests/tests.rs:25]: rustfmt wants to drop this newline: https://travis-ci.org/github/eqrion/cbindgen/builds/670071458

File: tests/tests.rs [hunk 3]
```diff
@@ -21,6 +22,10 @@ fn run_cbindgen(
...
+    let tmp_dir = TempDir::new("cbindgen").unwrap();
+    env::set_current_dir(tmp_dir.path()).unwrap();
```
Comment (emilio) [inline tests/tests.rs:27]: This doesn't look good, this can run in parallel for different tests.

Comment (emilio) [reply tests/tests.rs:27]: Also doesn't look like the right place. It should be in `compile()` instead.

File: tests/tests.rs [hunk 4]
```diff
@@ -121,7 +121,8 @@ fn run_compile_test(
     style: Option<Style>,
 ) {
     let crate_dir = env::var("CARGO_MANIFEST_DIR").unwrap();
-    let tests_path = Path::new(&crate_dir).join("tests");
+    let tmp_dir = TempDir::new("cbindgen").unwrap();
+    let tests_path = tmp_dir.path();
```
Comment (emilio) [inline tests/tests.rs:125]: This doesn't make sense to me. This breaks updating the test expectations and so on.

Comment (emilio) [reply tests/tests.rs:125]: Ah, I lie, but this still breaks tests that expect to find includes in `tests_path`. Why is this one needed.

**Comment by emilio** [comment]
I tweaked this (we had another tmpdir dependency already) and landed it in https://github.com/eqrion/cbindgen/commit/39bae60ad7ac4711d8fccc0d098ecef98b37d559. Thank you!

**Comment by sylvestre** [comment]
@emilio are you sure?
I tried to cherry pick your patch in 0.13.2 and it failed with

`thread 'test_va_list' panicked at 'Output failed to compile: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "Assembler messages:\nFatal error: can\'t create /usr/share/cargo/registry/cbindgen-0.13.2/tests/expectations/va_list.compat.o: Permission denied\n" }', tests/tests.rs:112:5
`

**Comment by sylvestre** [comment]
to reproduce:
 chmod  -w $(fd tests)

**Comment by emilio** [comment]
I cannot repro that panic with those str

**Comment by emilio** [comment]
ah, I know what goes wrong. But in fairness it's something that was wrong with your patch already :)

**Comment by sylvestre** [comment]
yeah, my patch was wrong too :)

**Comment by emilio** [comment]
a519f1b should fix that.

**Pull request closed without merging**